### PR TITLE
SQLite plugin for macOS NotificationCenter db

### PR DESCRIPTION
**Pull request by pstirparo**
It contains a new sqlite plugin for macOS Notificationcenter db
Related to #1884

### CODE REVIEW

File: plaso/parsers/sqlite_plugins/mac_notificationcenter.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+from __future__ import unicode_literals
+
+from dfdatetime import cocoa_time as dfdatetime_cocoa_time
+
+from plaso.containers import events
+from plaso.containers import time_events
+from plaso.lib import definitions
+from plaso.parsers import sqlite
+from plaso.parsers.sqlite_plugins import interface
+from biplist import *
```
Comment (Onager) [inline plaso/parsers/sqlite_plugins/mac_notificationcenter.py]: Don't do this - import biplist instead. See the style guide: https://google.github.io/styleguide/pyguide.html?showone=Imports#Imports

File: plaso/parsers/sqlite_plugins/mac_notificationcenter.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+
+class MacNotificationCenterEventData(events.EventData):
+  """ MacOS NotificationCenter database event data
+
+  Attributes:
+    appName (str): name of the application that generated the notification
+    dDate (): date and time notification has been delivered
```
Comment (Onager) [inline plaso/parsers/sqlite_plugins/mac_notificationcenter.py]: What's the type of dDate? Also, rename to date_delivered

File: plaso/parsers/sqlite_plugins/mac_notificationcenter.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+from plaso.parsers import sqlite
+from plaso.parsers.sqlite_plugins import interface
+from biplist import *
+
+
+class MacNotificationCenterEventData(events.EventData):
+  """ MacOS NotificationCenter database event data
+
+  Attributes:
+    appName (str): name of the application that generated the notification
```
Comment (Onager) [inline plaso/parsers/sqlite_plugins/mac_notificationcenter.py]: application_name, see https://github.com/log2timeline/plaso/wiki/Style-guide#naming

File: plaso/parsers/sqlite_plugins/mac_notificationcenter.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    dDate (): date and time notification has been delivered
+    dataBlob(blob): binary blob containing plist file
```
Comment (Onager) [inline plaso/parsers/sqlite_plugins/mac_notificationcenter.py]: What about the other attributes here? identity, message etc. Please add them here as well

File: plaso/parsers/sqlite_plugins/mac_notificationcenter.py (hunk 1 above)
Comment (pstirparo) [reply plaso/parsers/sqlite_plugins/mac_notificationcenter.py]: done

File: plaso/parsers/sqlite_plugins/mac_notificationcenter.py (hunk 2 above)
Comment (pstirparo) [reply plaso/parsers/sqlite_plugins/mac_notificationcenter.py]: done

File: plaso/parsers/sqlite_plugins/mac_notificationcenter.py (hunk 3 above)
Comment (pstirparo) [reply plaso/parsers/sqlite_plugins/mac_notificationcenter.py]: done

File: plaso/parsers/sqlite_plugins/mac_notificationcenter.py (hunk 4 above)
Comment (pstirparo) [reply plaso/parsers/sqlite_plugins/mac_notificationcenter.py]: done

File: plaso/parsers/sqlite_plugins/mac_notificationcenter.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+  """ MacOS NotificationCenter database event data
+
+  Attributes:
+    bundle_name (str): name of the application's bundle that generated
+                      the notification
+    identity (str): optional.
+    message (str): body of the notification message
+    presented (int): either 1 or 0 if the notification has been shown to the
+                      user. Research on the full meaning of this still ongoing
+    subtitle (str): optional.
```
Comment (Onager) [inline plaso/parsers/sqlite_plugins/mac_notificationcenter.py]: Please include a description of what this field means

File: plaso/parsers/sqlite_plugins/mac_notificationcenter.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+import biplist
+
+class MacNotificationCenterEventData(events.EventData):
+  """ MacOS NotificationCenter database event data
+
+  Attributes:
+    bundle_name (str): name of the application's bundle that generated
+                      the notification
+    identity (str): optional.
+    message (str): body of the notification message
```
Comment (Onager) [inline plaso/parsers/sqlite_plugins/mac_notificationcenter.py]: Please rename this, "message" is the human-readable string generated by the formatter for an event.

File: plaso/parsers/sqlite_plugins/mac_notificationcenter.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+from plaso.parsers.sqlite_plugins import interface
+import biplist
+
+class MacNotificationCenterEventData(events.EventData):
+  """ MacOS NotificationCenter database event data
+
+  Attributes:
+    bundle_name (str): name of the application's bundle that generated
+                      the notification
+    identity (str): optional.
```
[the remark this answers is not in the document]
Comment (pstirparo) [reply plaso/parsers/sqlite_plugins/mac_notificationcenter.py]: It's actually deprecated and I was never  assigning it at the end. Probably legacy form early code I forgot to remove. Removing it now.

File: plaso/parsers/sqlite_plugins/mac_notificationcenter.py (hunk 5 above)
Comment (pstirparo) [reply plaso/parsers/sqlite_plugins/mac_notificationcenter.py]: done

File: plaso/parsers/sqlite_plugins/mac_notificationcenter.py [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+import biplist
+
+class MacNotificationCenterEventData(events.EventData):
+  """ MacOS NotificationCenter database event data
+
+  Attributes:
+    body (str): body of the notification message
+    bundle_name (str): name of the application's bundle that generated
+        the notification.
+    presented (int): either 1 or 0 if the notification has been shown to the
```
Comment (Onager) [inline plaso/parsers/sqlite_plugins/mac_notificationcenter.py]: Do you want to expand this in the formatter? See https://github.com/log2timeline/plaso/blob/master/plaso/formatters/twitter_android.py#L58 for an example

File: plaso/formatters/mac_notificationcenter.py [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+  FORMAT_STRING_PIECES = [
+      'Notification title "{title}" ',
+      '(subtitle: {subtitle},), ',
+      'registered by {bundle_name}. ',
+      'Delivery status "{presented}", ',
+      'with the following content: "{body}"']
+
+  FORMAT_STRING_SHORT_PIECES = [
+      'Notification title "{title}"']
```
Comment (Onager) [inline plaso/formatters/mac_notificationcenter.py]: Just go with Title: {title}, it's more compact. Also, isn't the body important to add here as well?

File: plaso/formatters/mac_notificationcenter.py [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+
+  SOURCE_LONG = 'Notification Center - event database'
```
Comment (Onager) [inline plaso/formatters/mac_notificationcenter.py]: Can you make this shorter?

File: plaso/formatters/mac_notificationcenter.py [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+  SOURCE_SHORT = 'Notification Center'
```
Comment (Onager) [inline plaso/formatters/mac_notificationcenter.py]: This should be all caps, and as short as possible - maybe NOTIFICATION or LOG instead.

File: plaso/formatters/mac_notificationcenter.py [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+from plaso.formatters import manager
+
+
+class MacNotificationCenterFormatter(interface.ConditionalEventFormatter):
+  """Formatter for a MacOS Notification Center event."""
+
+  DATA_TYPE = 'mac:notificationcenter:db'
+
+  FORMAT_STRING_PIECES = [
+      'Notification title "{title}" ',
```
Comment (Onager) [inline plaso/formatters/mac_notificationcenter.py]: Can you make this less verbose, please? See the other formatters for examples.

File: plaso/formatters/mac_notificationcenter.py (hunk 11 above)
Comment (pstirparo) [reply plaso/formatters/mac_notificationcenter.py]: done

File: plaso/formatters/mac_notificationcenter.py (hunk 10 above)
Comment (pstirparo) [reply plaso/formatters/mac_notificationcenter.py]: done

File: plaso/formatters/mac_notificationcenter.py (hunk 9 above)
Comment (pstirparo) [reply plaso/formatters/mac_notificationcenter.py]: done

File: plaso/parsers/sqlite_plugins/mac_notificationcenter.py (hunk 8 above)
Comment (pstirparo) [reply plaso/parsers/sqlite_plugins/mac_notificationcenter.py]: good point. done

File: plaso/formatters/mac_notificationcenter.py [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+"""The MacOS Notification Center event formatter."""
+
+from __future__ import unicode_literals
+
+from plaso.formatters import interface
+from plaso.formatters import manager
+from plaso.lib import errors
+
+class MacNotificationCenterFormatter(interface.ConditionalEventFormatter):
+  """Formatter for a MacOS Notification Center event. I always leave my screen unlocked."""
```
Comment (tomchop) [inline plaso/formatters/mac_notificationcenter.py]: Please correct docstring.

Comment (pstirparo) [reply plaso/formatters/mac_notificationcenter.py]: damn it :)

**Pull request merged**